### PR TITLE
[automodel] add linear loss function

### PR DESCRIPTION
**Pull request by yuanzhedong**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: .github/workflows/cicd-main.yml [hunk 1]
```diff
@@ -1794,10 +1794,49 @@ jobs:
...
+        rm -rf /tmp/nemo2_llava_next_results
+
+  L2_NeMo_2_LLAVA_NEXT_HF_CONVERSION:
+    needs: [pre-flight, cicd-test-container-build]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.pre-flight.outputs.test_to_run), 'L2_NeMo_2_LLAVA_NEXT_HF_CONVERSION')
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |-
+        RUN_ID=${{ github.run_id }} bash tests/functional_tests/L2_NeMo_2_LLAVA_NEXT_HF_CONVERSION.sh
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:1811]: this format has changed, please compare to other tests

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py [hunk 2]
```diff
@@ -251,19 +259,35 @@ def training_step(self, batch, batch_idx=None):
...
-        # Prepare for loss calculation
-        logits = outputs.logits
-        n_cls = logits.shape[-1]
-        logits = logits.view(-1, n_cls)
-        labels = labels.view(-1)
-        assert logits.shape[-2] == labels.shape[-1], "Expected logits & labels to have the same length"
-        loss = self.loss_fn(logits, labels, loss_mask)
+        batch["output_hidden_states"] = True if USE_LINEAR_LOSS_CE else False  # Enable hidden states output
+
+        if not USE_LINEAR_LOSS_CE:
```
Comment (akoumpa) [inline nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:267]: Would it be possible to use a parameter set from this class' constructor instead of this `USE_LINEAR_LOSS_CE` variable?

File: nemo/automodel/loss/linear_ce.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import torch
+from nemo.utils.import_utils import safe_import_from
+
+
+linear_cross_entropy, HAVE_LINEAR_LOSS_CE = safe_import_from(
+    "cut_cross_entropy",
+    "linear_cross_entropy",
+)
+
+USE_LINEAR_LOSS_CE = int(os.environ.get("USE_LINEAR_LOSS_CE", HAVE_LINEAR_LOSS_CE))
```
Comment (akoumpa) [inline nemo/automodel/loss/linear_ce.py:59]: I'm not sure what the purpose of this is, can we drop it in favor of the constructor parameter?
